lgc/patch/PatchBoundsCheckMemory.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchBoundsCheckMemory.cpp
+ * @brief LLPC source file: contains implementation of class lgc::PatchBoundsCheckMemory.
+ *
+ * This pass adds bounds checks to all stack/scratch accesses with dynamic indices. The pass looks at every
+ * getelementptr instruction, checking which of the used indices are not constants. For non-constant indices, it checks
+ * if they index into either a FixedVectorType or an ArrayType and gets their element count, which is the upper bound of
+ * the index.
+ *
+ * Example:
+ * We have a getelementptr, followed by a load.
+ *
+ * %elemPtr = getelementptr [16 x float], [16 x float] addrspace(5)* %array, i32 0, i32 %index
```

Comment:
Does you insert checks regardless of the GEP `inbounds` annotation?

Reply:
Yes